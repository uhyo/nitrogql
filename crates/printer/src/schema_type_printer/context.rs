use std::{
    borrow::Cow,
    collections::{HashMap, HashSet},
};

use graphql_type_system::Schema;
use nitrogql_ast::{
    base::Pos,
    type_system::{TypeDefinition, TypeSystemDefinition},
    TypeSystemDocument,
};
use nitrogql_config_file::{ScalarTypeConfig, SeparateScalarTypeConfig, TypeTarget};

use crate::SchemaTypePrinterOptions;

pub struct SchemaTypePrinterContext<'src> {
    pub options: &'src SchemaTypePrinterOptions,
    pub document: &'src TypeSystemDocument<'src>,
    pub schema: &'src Schema<Cow<'src, str>, Pos>,
    // Mapping from Scalar name to TypeScript types.
    pub scalar_types: HashMap<String, ScalarTypeConfig>,
    /// Mapping from schema type name to local type name.
    pub local_type_names: HashMap<String, String>,
    /// Current output type target.
    pub type_target: TypeTarget,
}

impl SchemaTypePrinterContext<'_> {
    pub fn new<'src>(
        options: &'src SchemaTypePrinterOptions,
        document: &'src TypeSystemDocument<'src>,
        schema: &'src Schema<Cow<'src, str>, Pos>,
        type_target: TypeTarget,
    ) -> SchemaTypePrinterContext<'src> {
        let scalar_types = get_scalar_types(document, options);
        let local_type_names = make_local_type_names(document, &scalar_types);
        SchemaTypePrinterContext {
            options,
            document,
            schema,
            scalar_types,
            local_type_names,
            type_target,
        }
    }
}

// Generates TS Types for all scalars.
fn get_scalar_types(
    document: &TypeSystemDocument,
    options: &SchemaTypePrinterOptions,
) -> HashMap<String, ScalarTypeConfig> {
    document
        .definitions
        .iter()
        .filter_map(|definition| match definition {
            TypeSystemDefinition::TypeDefinition(TypeDefinition::Scalar(definition)) => {
                Some(definition)
            }
            _ => None,
        })
        .filter_map(|definition| {
            // type of scalar has two sources:
            // @nitrogql_ts_type built-in directive and scalarTypes option.
            // If scalarType is provided, it takes precedence.
            let scalar_type_from_config = options.scalar_types.get(definition.name.name);
            let directive_ts_type = definition
                .directives
                .iter()
                .find(|directive| (directive.name.name == "nitrogql_ts_type"))
                .and_then(|directive| directive.arguments.as_ref())
                .and_then(|args| {
                    let mut resolver_input_type = None;
                    let mut resolver_output_type = None;
                    let mut operation_input_type = None;
                    let mut operation_output_type = None;
                    for (key, value) in args.arguments.iter() {
                        match key.name {
                            "resolverInput" => resolver_input_type = Some(value),
                            "resolverOutput" => resolver_output_type = Some(value),
                            "operationInput" => operation_input_type = Some(value),
                            "operationOutput" => operation_output_type = Some(value),
                            _ => {}
                        }
                    }
                    if let (
                        Some(resolver_input_type),
                        Some(resolver_output_type),
                        Some(operation_input_type),
                        Some(operation_output_type),
                    ) = (
                        resolver_input_type.and_then(|v| v.as_string()),
                        resolver_output_type.and_then(|v| v.as_string()),
                        operation_input_type.and_then(|v| v.as_string()),
                        operation_output_type.and_then(|v| v.as_string()),
                    ) {
                        Some(ScalarTypeConfig::Separate(SeparateScalarTypeConfig {
                            resolver_input: resolver_input_type.value.clone(),
                            resolver_output: resolver_output_type.value.clone(),
                            operation_input: operation_input_type.value.clone(),
                            operation_output: operation_output_type.value.clone(),
                        }))
                    } else {
                        None
                    }
                });
            let scalar_ts_type = scalar_type_from_config.cloned().or(directive_ts_type);
            scalar_ts_type.map(|ty| (definition.name.name.to_owned(), ty))
        })
        .collect()
}

fn get_bag_of_identifiers(scalar_types: &HashMap<String, ScalarTypeConfig>) -> HashSet<&str> {
    let mut result = vec![];
    for value in scalar_types.values().flat_map(|v| v.type_names()) {
        let mut start_index = 0;
        let mut in_identifier = false;
        for (index, c) in value.char_indices() {
            if !in_identifier {
                if c.is_ascii_alphabetic() || c == '_' {
                    in_identifier = true;
                    start_index = index;
                }
            } else if !c.is_ascii_alphanumeric() && c != '_' {
                // end of identifier
                result.push(&value[start_index..index]);
                in_identifier = false;
            }
        }
        if in_identifier {
            result.push(&value[start_index..]);
        }
    }
    result.into_iter().collect()
}

fn make_local_type_names(
    document: &TypeSystemDocument,
    scalar_types: &HashMap<String, ScalarTypeConfig>,
) -> HashMap<String, String> {
    // The bag is the set of identifiers that appear in TypeScript types of scalars.
    // We will use this to avoid name collisions.
    let bag = get_bag_of_identifiers(scalar_types);
    document
        .definitions
        .iter()
        .filter_map(|definition| match definition {
            TypeSystemDefinition::TypeDefinition(def) => Some(def),
            _ => None,
        })
        .map(|definition| {
            let schema_name = definition.name().name;
            let local_name = if bag.contains(schema_name) {
                format!("__tmp_{schema_name}")
            } else {
                schema_name.to_owned()
            };
            (schema_name.to_owned(), local_name)
        })
        .collect()
}
