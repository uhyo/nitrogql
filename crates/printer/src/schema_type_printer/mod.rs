pub mod context;
mod error;
pub mod printer;
mod tests;
mod type_printer;
