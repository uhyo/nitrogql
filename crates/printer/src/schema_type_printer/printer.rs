use std::{borrow::Cow, collections::HashMap};

use graphql_type_system::Schema;
use nitrogql_ast::{base::Pos, type_system::TypeSystemDocument};
use nitrogql_config_file::Config;
use nitrogql_semantics::ast_to_type_system;
use sourcemap_writer::SourceMapWriter;

use crate::schema::get_builtin_scalar_types;

use super::{error::SchemaTypePrinterResult, type_printer::TypePrinter};

pub struct SchemaTypePrinterOptions {
    /// Type of each scalar. Provided as raw TypeScript code.
    pub scalar_types: HashMap<String, String>,
    /// Special type name for types that includes schema metadata
    pub schema_metadata_type: String,
    /// Whether to make input nullable fields optional.
    pub input_nullable_field_is_optional: bool,
    /// Whether to emit runtime for generated schema types.
    pub emit_schema_runtime: bool,
}

impl Default for SchemaTypePrinterOptions {
    fn default() -> Self {
        SchemaTypePrinterOptions {
            scalar_types: get_builtin_scalar_types(),
            schema_metadata_type: "__nitrogql_schema".into(),
            input_nullable_field_is_optional: true,
            emit_schema_runtime: false,
        }
    }
}

impl SchemaTypePrinterOptions {
    /// Generate from config.
    pub fn from_config(config: &Config) -> Self {
        let mut result = SchemaTypePrinterOptions {
            emit_schema_runtime: config.generate.emit_schema_runtime,
            input_nullable_field_is_optional: config
                .generate
                .r#type
                .allow_undefined_as_optional_input,
            ..SchemaTypePrinterOptions::default()
        };
        result.scalar_types.extend(
            config
                .generate
                .r#type
                .scalar_types
                .iter()
                .map(|(key, value)| (key.to_owned(), value.to_owned())),
        );
        result
    }
}

pub struct SchemaTypePrinterContext<'src> {
    pub options: &'src SchemaTypePrinterOptions,
    pub document: &'src TypeSystemDocument<'src>,
    pub schema: &'src Schema<Cow<'src, str>, Pos>,
}

pub struct SchemaTypePrinter<'a, Writer: SourceMapWriter> {
    options: SchemaTypePrinterOptions,
    writer: &'a mut Writer,
}

impl<'a, Writer> SchemaTypePrinter<'a, Writer>
where
    Writer: SourceMapWriter,
{
    pub fn new(options: SchemaTypePrinterOptions, writer: &'a mut Writer) -> Self {
        SchemaTypePrinter { options, writer }
    }

    pub fn print_document(&mut self, document: &TypeSystemDocument) -> SchemaTypePrinterResult<()> {
        let schema = ast_to_type_system(document);
        let context = SchemaTypePrinterContext {
            options: &self.options,
            document,
            schema: &schema,
        };
        document.print_type(&context, self.writer)?;
        Ok(())
    }
}
