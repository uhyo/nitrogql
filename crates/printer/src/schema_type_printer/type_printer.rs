use std::{borrow::Borrow, fmt::Display};

use crate::{
    ts_types::{
        ts_types_util::ts_union, type_to_ts_type::get_ts_type_of_type, ObjectField, TSType,
    },
    utils::interface_implementers,
};
use nitrogql_ast::{
    base::{HasPos, Ident, Pos},
    type_system::{
        EnumTypeDefinition, InputObjectTypeDefinition, InterfaceTypeDefinition,
        ObjectTypeDefinition, ScalarTypeDefinition, TypeDefinition, TypeSystemDefinition,
        UnionTypeDefinition,
    },
    value::StringValue,
};
use nitrogql_config_file::TypeTarget;
use sourcemap_writer::SourceMapWriter;

use crate::jsdoc::print_description as jsdoc_print_description;

use super::{
    context::SchemaTypePrinterContext,
    error::{SchemaTypePrinterError, SchemaTypePrinterResult},
};

pub trait TypePrinter {
    fn print_type(
        &self,
        context: &SchemaTypePrinterContext,
        writer: &mut impl SourceMapWriter,
    ) -> SchemaTypePrinterResult<()>;
    fn print_representative(
        &self,
        context: &SchemaTypePrinterContext,
        writer: &mut impl SourceMapWriter,
    ) -> SchemaTypePrinterResult<()>;
}

impl TypePrinter for TypeSystemDefinition<'_> {
    fn print_type(
        &self,
        context: &SchemaTypePrinterContext,
        writer: &mut impl SourceMapWriter,
    ) -> SchemaTypePrinterResult<()> {
        match self {
            TypeSystemDefinition::SchemaDefinition(_) => Ok(()),
            TypeSystemDefinition::TypeDefinition(def) => def.print_type(context, writer),
            TypeSystemDefinition::DirectiveDefinition(_) => Ok(()),
        }
    }
    fn print_representative(
        &self,
        context: &SchemaTypePrinterContext,
        writer: &mut impl SourceMapWriter,
    ) -> SchemaTypePrinterResult<()> {
        match self {
            TypeSystemDefinition::SchemaDefinition(_) => Ok(()),
            TypeSystemDefinition::TypeDefinition(def) => def.print_representative(context, writer),
            TypeSystemDefinition::DirectiveDefinition(_) => Ok(()),
        }
    }
}

impl TypePrinter for TypeDefinition<'_> {
    fn print_type(
        &self,
        context: &SchemaTypePrinterContext,
        writer: &mut impl SourceMapWriter,
    ) -> SchemaTypePrinterResult<()> {
        match self {
            TypeDefinition::Scalar(def) => def.print_type(context, writer),
            TypeDefinition::Object(def) => def.print_type(context, writer),
            TypeDefinition::Interface(def) => def.print_type(context, writer),
            TypeDefinition::Union(def) => def.print_type(context, writer),
            TypeDefinition::Enum(def) => def.print_type(context, writer),
            TypeDefinition::InputObject(def) => def.print_type(context, writer),
        }
    }
    fn print_representative(
        &self,
        context: &SchemaTypePrinterContext,
        writer: &mut impl SourceMapWriter,
    ) -> SchemaTypePrinterResult<()> {
        match self {
            TypeDefinition::Scalar(def) => def.print_representative(context, writer),
            TypeDefinition::Object(def) => def.print_representative(context, writer),
            TypeDefinition::Interface(def) => def.print_representative(context, writer),
            TypeDefinition::Union(def) => def.print_representative(context, writer),
            TypeDefinition::Enum(def) => def.print_representative(context, writer),
            TypeDefinition::InputObject(def) => def.print_representative(context, writer),
        }
    }
}

impl TypePrinter for ScalarTypeDefinition<'_> {
    fn print_type(
        &self,
        context: &SchemaTypePrinterContext,
        writer: &mut impl SourceMapWriter,
    ) -> SchemaTypePrinterResult<()> {
        let scalar_ts_type = context.scalar_types.get(self.name.name);
        let Some(scalar_type_str) = scalar_ts_type else {
            return Err(SchemaTypePrinterError::ScalarTypeNotProvided {
                position: self.position,
                name: self.name.to_string(),
            });
        };

        print_description(&self.description, writer);
        let local_name = context
            .local_type_names
            .get(self.name.name)
            .expect("Local type name not generated");

        export_type(
            writer,
            &self.scalar_keyword,
            &self.name,
            local_name,
            |writer| {
                writer.write(scalar_type_str.get_type(context.type_target));
            },
        );
        Ok(())
    }
    fn print_representative(
        &self,
        context: &SchemaTypePrinterContext,
        writer: &mut impl SourceMapWriter,
    ) -> SchemaTypePrinterResult<()> {
        let local_name = context
            .local_type_names
            .get(self.name.name)
            .expect("Local type name not generated");
        export_representative(
            writer,
            &self.scalar_keyword,
            &self.name,
            local_name,
            TypeTarget::OperationOutput,
        );
        Ok(())
    }
}

impl TypePrinter for ObjectTypeDefinition<'_> {
    fn print_type(
        &self,
        context: &SchemaTypePrinterContext,
        writer: &mut impl SourceMapWriter,
    ) -> SchemaTypePrinterResult<()> {
        if context.type_target.is_input() {
            // Object type is not used as input type.
            return Ok(());
        }
        let type_name_ident = Ident {
            name: "__typename",
            position: Pos::builtin(),
        };
        let schema_type = context.schema.get_type(self.name.name);
        let obj_type = TSType::object(
            vec![(
                &type_name_ident,
                TSType::StringLiteral(self.name.to_string()),
                None,
            )]
            .into_iter()
            .chain(self.fields.iter().map(|field| {
                let schema_field = schema_type
                    .and_then(|ty| ty.as_object())
                    .and_then(|ty| ty.fields.iter().find(|f| f.name == field.name.name));
                (
                    &field.name,
                    get_ts_type_of_type(&field.r#type, |name| {
                        let local_name = context
                            .local_type_names
                            .get(name.name.name)
                            .expect("Local type name not generated");
                        TSType::TypeVariable(local_name.as_str().into())
                    }),
                    make_ts_description(
                        &field.description,
                        &schema_field.and_then(|f| f.deprecation.as_ref()),
                    ),
                )
            })),
        );

        print_description(&self.description, writer);
        let local_name = context
            .local_type_names
            .get(self.name.name)
            .expect("Local type name not generated");

        export_type(
            writer,
            &self.type_keyword,
            &self.name,
            local_name,
            |writer| {
                obj_type.print_type(writer);
            },
        );
        Ok(())
    }
    fn print_representative(
        &self,
        context: &SchemaTypePrinterContext,
        writer: &mut impl SourceMapWriter,
    ) -> SchemaTypePrinterResult<()> {
        let local_name = context
            .local_type_names
            .get(self.name.name)
            .expect("Local type name not generated");
        export_representative(
            writer,
            &self.type_keyword,
            &self.name,
            local_name,
            TypeTarget::OperationOutput,
        );
        Ok(())
    }
}

impl TypePrinter for InterfaceTypeDefinition<'_> {
    fn print_type(
        &self,
        context: &SchemaTypePrinterContext,
        writer: &mut impl SourceMapWriter,
    ) -> SchemaTypePrinterResult<()> {
        if context.type_target.is_input() {
            // Interface is not used as input type.
            return Ok(());
        }
        // In generated type definitions, an interface is expressed as a union of all possible concrete types.
        let union_constituents =
            interface_implementers(context.schema, self.name.name).map(|obj| {
                TSType::TypeVariable({
                    let s: &str = (obj.name).inner_ref().borrow();
                    s.into()
                })
            });
        let intf_type = ts_union(union_constituents);

        print_description(&self.description, writer);
        let local_name = context
            .local_type_names
            .get(self.name.name)
            .expect("Local type name not generated");
        export_type(
            writer,
            &self.interface_keyword,
            &self.name,
            local_name,
            |writer| {
                intf_type.print_type(writer);
            },
        );
        Ok(())
    }
    fn print_representative(
        &self,
        context: &SchemaTypePrinterContext,
        writer: &mut impl SourceMapWriter,
    ) -> SchemaTypePrinterResult<()> {
        let local_name = context
            .local_type_names
            .get(self.name.name)
            .expect("Local type name not generated");
        export_representative(
            writer,
            &self.interface_keyword,
            &self.name,
            local_name,
            TypeTarget::OperationOutput,
        );
        Ok(())
    }
}

impl TypePrinter for UnionTypeDefinition<'_> {
    fn print_type(
        &self,
        context: &SchemaTypePrinterContext,
        writer: &mut impl SourceMapWriter,
    ) -> SchemaTypePrinterResult<()> {
        if context.type_target.is_input() {
            // Union is not used as input type.
            return Ok(());
        }
        let union_type = ts_union(
            self.members
                .iter()
                .map(|mem| TSType::TypeVariable(mem.into())),
        );

        print_description(&self.description, writer);
        let local_name = context
            .local_type_names
            .get(self.name.name)
            .expect("Local type name not generated");
        export_type(
            writer,
            &self.union_keyword,
            &self.name,
            local_name,
            |writer| {
                union_type.print_type(writer);
            },
        );
        Ok(())
    }
    fn print_representative(
        &self,
        context: &SchemaTypePrinterContext,
        writer: &mut impl SourceMapWriter,
    ) -> SchemaTypePrinterResult<()> {
        let local_name = context
            .local_type_names
            .get(self.name.name)
            .expect("Local type name not generated");
        export_representative(
            writer,
            &self.union_keyword,
            &self.name,
            local_name,
            TypeTarget::OperationOutput,
        );
        Ok(())
    }
}

impl TypePrinter for EnumTypeDefinition<'_> {
    fn print_type(
        &self,
        context: &SchemaTypePrinterContext,
        writer: &mut impl SourceMapWriter,
    ) -> SchemaTypePrinterResult<()> {
        let enum_type = TSType::Union(
            self.values
                .iter()
                .map(|mem| TSType::StringLiteral(mem.name.to_string()))
                .collect(),
        );

        print_description(&self.description, writer);
        let local_name = context
            .local_type_names
            .get(self.name.name)
            .expect("Local type name not generated");
        export_type(
            writer,
            &self.enum_keyword,
            &self.name,
            local_name,
            |writer| {
                enum_type.print_type(writer);
            },
        );
        Ok(())
    }
    fn print_representative(
        &self,
        context: &SchemaTypePrinterContext,
        writer: &mut impl SourceMapWriter,
    ) -> SchemaTypePrinterResult<()> {
        let local_name = context
            .local_type_names
            .get(self.name.name)
            .expect("Local type name not generated");
        export_representative(
            writer,
            &self.enum_keyword,
            &self.name,
            local_name,
            TypeTarget::OperationOutput,
        );
        if context.options.emit_schema_runtime {
            writer.write_for("export const ", &self.enum_keyword);
            writer.write_for(self.name.name, &self.name);
            writer.write(" = {\n");
            writer.indent();
            for value in &self.values {
                writer.write_for(value.name.name, &value.name);
                writer.write(": \"");
                writer.write_for(value.name.name, &value.name);
                writer.write("\",\n");
            }
            writer.dedent();
            writer.write("} as const;\n")
        }
        Ok(())
    }
}

impl TypePrinter for InputObjectTypeDefinition<'_> {
    fn print_type(
        &self,
        context: &SchemaTypePrinterContext,
        writer: &mut impl SourceMapWriter,
    ) -> SchemaTypePrinterResult<()> {
        if context.type_target.is_output() {
            // Input object is not used as output type.
            return Ok(());
        }
        let schema_type = context.schema.get_type(self.name.name);
        let obj_type = TSType::Object(
            self.fields
                .iter()
                .map(|field| {
                    let schema_field = schema_type
                        .and_then(|t| t.as_input_object())
                        .and_then(|t| t.fields.iter().find(|f| f.name == field.name.name))
                        .expect("Type system error");

                    let ts_type = get_ts_type_of_type(&field.r#type, |name| {
                        let local_name = context
                            .local_type_names
                            .get(name.name.name)
                            .expect("Local type name not generated");
                        TSType::TypeVariable(local_name.as_str().into())
                    })
                    .into_readonly();
                    let is_optional = context.options.input_nullable_field_is_optional
                        && !field.r#type.is_nonnull();
                    let ts_type = if is_optional {
                        TSType::Union(vec![ts_type, TSType::Undefined])
                    } else {
                        ts_type
                    };
                    ObjectField {
                        key: (&field.name).into(),
                        r#type: ts_type,
                        readonly: true,
                        optional: is_optional,
                        description: make_ts_description(
                            &field.description,
                            &schema_field.deprecation,
                        ),
                    }
                })
                .collect(),
        );

        print_description(&self.description, writer);
        let local_name = context
            .local_type_names
            .get(self.name.name)
            .expect("Local type name not generated");
        export_type(
            writer,
            &self.input_keyword,
            &self.name,
            local_name,
            |writer| {
                obj_type.print_type(writer);
            },
        );
        Ok(())
    }
    fn print_representative(
        &self,
        context: &SchemaTypePrinterContext,
        writer: &mut impl SourceMapWriter,
    ) -> SchemaTypePrinterResult<()> {
        let local_name = context
            .local_type_names
            .get(self.name.name)
            .expect("Local type name not generated");
        export_representative(
            writer,
            &self.input_keyword,
            &self.name,
            local_name,
            TypeTarget::ResolverInput,
        );
        Ok(())
    }
}

fn export_type<Writer: SourceMapWriter>(
    writer: &mut Writer,
    type_keyword: &impl HasPos,
    schema_name: &Ident,
    local_name: &str,
    print_type: impl FnOnce(&mut Writer),
) {
    if schema_name.name == local_name {
        writer.write_for("export type ", type_keyword);
        writer.write_for(local_name, schema_name);
        writer.write(" = ");
        print_type(writer);
        writer.write(";\n");
    } else {
        writer.write_for("type ", type_keyword);
        writer.write_for(local_name, schema_name);
        writer.write(" = ");
        print_type(writer);
        writer.write(";\nexport type { ");
        writer.write(local_name);
        writer.write(" as ");
        writer.write(schema_name.name);
        writer.write("};\n");
    }
}

fn export_representative(
    writer: &mut impl SourceMapWriter,
    type_keyword: &impl HasPos,
    schema_name: &Ident,
    local_name: &str,
    target: TypeTarget,
) {
    if schema_name.name == local_name {
        writer.write_for("export type ", type_keyword);
        writer.write_for(local_name, schema_name);
        writeln!(writer, " = {target}.{local_name};");
    } else {
        writer.write_for("type ", type_keyword);
        writer.write_for(local_name, schema_name);
        writeln!(writer, " = {target}.{schema_name};");
        writeln!(
            writer,
            "export type {{ {local_name} as {} }};",
            schema_name.name
        );
    }
}

fn print_description(description: &Option<StringValue>, writer: &mut impl SourceMapWriter) {
    if let Some(description) = description {
        jsdoc_print_description(description, writer);
    }
}

/// Combines description and deprecation reason into a single string.
fn make_ts_description(
    description: &Option<StringValue>,
    deprecation: &Option<impl Display>,
) -> Option<String> {
    match (description, deprecation) {
        (Some(description), Some(deprecation)) => Some(format!(
            "{}\n\n@deprecated {}",
            description.value, deprecation
        )),
        (Some(description), None) => Some(description.value.clone()),
        (None, Some(deprecation)) => format!("@deprecated {}", deprecation).into(),
        (None, None) => None,
    }
}
