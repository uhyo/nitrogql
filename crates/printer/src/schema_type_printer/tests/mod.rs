#![cfg(test)]

use insta::assert_snapshot;
use nitrogql_ast::type_system::TypeSystemDocument;

use crate::schema_type_printer::{
    error::SchemaTypePrinterResult,
    printer::{SchemaTypePrinter, SchemaTypePrinterOptions},
};
use nitrogql_parser::parse_type_system_document;
use nitrogql_semantics::resolve_schema_extensions;
use sourcemap_writer::JustWriter;

#[test]
fn type_printing() {
    let doc = parse_type_system_document(
        "
            scalar ID
            scalar String
            scalar Int

            type User implements HasID {
                id: ID!
                name: String!
                type: UserType!
                age: Int
                posts: [Post!]!
            }
            interface HasID {
                id: ID!
            }
            enum UserType {
                NormalUser
                PremiumUser
            }
            type Bot implements HasID {
                id: ID!
            }

            type Post {
                id: ID!
                title: String!
                tags: [String!]
                body: String!
            }

            input UserSearchQuery {
                age: Int
                name: String
                keywords: [String!]
            }

            type Query {
                me: User!
            }
            ",
    )
    .unwrap();
    let doc = resolve_schema_extensions(doc).unwrap();
    let printed = print_document(&doc, Default::default()).unwrap();
    assert_snapshot!(printed);
}

#[test]
fn type_printing_with_desc() {
    let doc = parse_type_system_document(
        r#"
            scalar ID
            scalar String
            scalar Int

            "This is User."
            type User implements HasID {
                id: ID!
                "Name of user."
                name: String!
                type: UserType!
                "Age of user. User may choose to not register their age."
                age: Int
                posts: [Post!]!
            }
            """
            Node that has an id field.
            """
            interface HasID {
                "ID of node that is globally unique."
                id: ID!
            }
            enum UserType {
                "Stands for free plan users."
                NormalUser
                "Stands for paid plan users."
                PremiumUser
            }
            type Bot implements HasID {
                id: ID!
            }

            type Post {
                id: ID!
                title: String!
                tags: [String!]
                body: String!
            }

            input UserSearchQuery {
                age: Int
                name: String
            }

            type Query {
                """
                Returns my account.
                Note that query without authorization header results in a error.
                """
                me: User!
            }
            "#,
    )
    .unwrap();
    let doc = resolve_schema_extensions(doc).unwrap();
    let printed = print_document(&doc, Default::default()).unwrap();
    assert_snapshot!(printed);
}

#[test]
fn scalar_printing() {
    let doc = parse_type_system_document(
        "
        scalar BigInt
        scalar URL
        ",
    )
    .unwrap();
    let doc = resolve_schema_extensions(doc).unwrap();
    let mut options = SchemaTypePrinterOptions::default();
    options.scalar_types.extend(vec![
        ("BigInt".to_owned(), "bigint".to_owned()),
        ("URL".to_owned(), "string".to_owned()),
    ]);
    let printed = print_document(&doc, options).unwrap();
    assert_snapshot!(printed);
}

#[test]
fn avoid_circular_reference_1() {
    let doc = parse_type_system_document(
        "
        scalar BigInt
        scalar Date
        type Obj {
            bigint: BigInt
            date: Date!
        }
        ",
    )
    .unwrap();
    let doc = resolve_schema_extensions(doc).unwrap();
    let mut options = SchemaTypePrinterOptions::default();
    options.scalar_types.extend(vec![
        ("BigInt".to_owned(), "bigint".to_owned()),
        ("Date".to_owned(), "Date".to_owned()),
    ]);
    let printed = print_document(&doc, options).unwrap();
    // Date should be emitted as __tmp_Date
    assert_snapshot!(printed);
}

#[test]
fn avoid_circular_reference_2() {
    let doc = parse_type_system_document(
        "
        scalar BigInt
        scalar Date
        type Obj {
            bigint: BigInt
            date: Date!
        }
        ",
    )
    .unwrap();
    let doc = resolve_schema_extensions(doc).unwrap();
    let mut options = SchemaTypePrinterOptions::default();
    options.scalar_types.extend(vec![
        ("BigInt".to_owned(), "bigint".to_owned()),
        ("Date".to_owned(), "Date | string".to_owned()),
    ]);
    let printed = print_document(&doc, options).unwrap();
    // Date should be emitted as __tmp_Date
    assert_snapshot!(printed);
}

#[test]
fn avoid_circular_reference_3() {
    let doc = parse_type_system_document(
        "
        scalar Rec1
        scalar Rec2
        type Obj {
            rec1: Rec1!
            rec2: Rec2!
        }
        ",
    )
    .unwrap();
    let doc = resolve_schema_extensions(doc).unwrap();
    let mut options = SchemaTypePrinterOptions::default();
    options.scalar_types.extend(vec![
        ("Rec1".to_owned(), "string | Rec2".to_owned()),
        ("Rec2".to_owned(), "Rec1 | number".to_owned()),
    ]);
    let printed = print_document(&doc, options).unwrap();
    assert_snapshot!(printed);
}

#[test]
fn deprecated_items() {
    let doc = parse_type_system_document(
        r#"
        scalar ID
        scalar String
        scalar Int

        type User {
            id: ID!
            name: String!
            "Age of user."
            age: Int @deprecated
            gender: String @deprecated(reason: "Deprecated for political reasons")
        }

        input UserSearchQuery {
            age: Int @deprecated
            name: String
        }

        type Query {
            me: User!
        }
        "#,
    )
    .unwrap();
    let doc = resolve_schema_extensions(doc).unwrap();
    let options = SchemaTypePrinterOptions::default();
    let printed = print_document(&doc, options).unwrap();
    assert_snapshot!(printed);
}

#[test]
fn enum_runtime() {
    let doc = parse_type_system_document(
        r#"
        enum UserType {
            NormalUser
            PremiumUser
            AdminUser
        }
        "#,
    )
    .unwrap();
    let doc = resolve_schema_extensions(doc).unwrap();
    let options = SchemaTypePrinterOptions {
        emit_schema_runtime: true,
        ..SchemaTypePrinterOptions::default()
    };
    let printed = print_document(&doc, options).unwrap();
    assert_snapshot!(printed);
}

fn print_document(
    document: &TypeSystemDocument,
    options: SchemaTypePrinterOptions,
) -> SchemaTypePrinterResult<String> {
    let mut result = String::new();
    let mut writer = JustWriter::new(&mut result);
    let mut printer = SchemaTypePrinter::new(options, &mut writer);
    printer.print_document(document)?;
    Ok(result)
}
