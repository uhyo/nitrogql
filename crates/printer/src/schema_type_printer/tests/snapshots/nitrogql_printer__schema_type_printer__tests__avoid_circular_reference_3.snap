---
source: crates/printer/src/schema_type_printer/tests/mod.rs
expression: printed
---
export type __nitrogql_schema = {};

type __Beautify<Obj> = { [K in keyof Obj]: Obj[K] } & {};
export type __SelectionSet<Orig, Obj, Others> =
  __Beautify<Pick<{
    [K in keyof Orig]: Obj extends { [P in K]?: infer V } ? V : unknown
  }, Extract<keyof Orig, keyof Obj>> & Others>;

export declare namespace __OperationInput {
  type __tmp_Rec1 = string | Rec2;
  export type { __tmp_Rec1 as Rec1};

  type __tmp_Rec2 = Rec1 | number;
  export type { __tmp_Rec2 as Rec2};

  export type Obj = {
    __typename: "Obj";
    rec1: __tmp_Rec1;
    rec2: __tmp_Rec2;
  };

}

export declare namespace __OperationOutput {
  type __tmp_Rec1 = string | Rec2;
  export type { __tmp_Rec1 as Rec1};

  type __tmp_Rec2 = Rec1 | number;
  export type { __tmp_Rec2 as Rec2};

  export type Obj = {
    __typename: "Obj";
    rec1: __tmp_Rec1;
    rec2: __tmp_Rec2;
  };

}

export declare namespace __ResolverInput {
  type __tmp_Rec1 = string | Rec2;
  export type { __tmp_Rec1 as Rec1};

  type __tmp_Rec2 = Rec1 | number;
  export type { __tmp_Rec2 as Rec2};

  export type Obj = {
    __typename: "Obj";
    rec1: __tmp_Rec1;
    rec2: __tmp_Rec2;
  };

}

export declare namespace __ResolverOutput {
  type __tmp_Rec1 = string | Rec2;
  export type { __tmp_Rec1 as Rec1};

  type __tmp_Rec2 = Rec1 | number;
  export type { __tmp_Rec2 as Rec2};

  export type Obj = {
    __typename: "Obj";
    rec1: __tmp_Rec1;
    rec2: __tmp_Rec2;
  };

}

type __tmp_Rec1 = __OperationOutput.Rec1;
export type { __tmp_Rec1 as Rec1 };

type __tmp_Rec2 = __OperationOutput.Rec2;
export type { __tmp_Rec2 as Rec2 };

export type Obj = __OperationOutput.Obj;


